This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [1]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
API_KEY = '' # removed before posting to Github

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [2]:
pip install requests

Note: you may need to restart the kernel to use updated packages.


In [3]:
# First, import the relevant modules
import requests
import collections
import json

In [4]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
url = 'https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?start_date=2020-09-30&end_date=2020-09-30&api_key=t8sGXjPhKLC1vLVUsUsn' 
r = requests.get(url)

In [5]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
text = r.text
print(text)

{"dataset":{"id":10095370,"dataset_code":"AFX_X","database_code":"FSE","name":"Carl Zeiss Meditec (AFX_X)","description":"Stock Prices for Carl Zeiss Meditec (2020-09-22) from the Frankfurt Stock Exchange.\u003cbr\u003e\u003cbr\u003eTrading System: Xetra\u003cbr\u003e\u003cbr\u003eISIN: DE0005313704","refreshed_at":"2020-10-06T22:16:16.250Z","newest_available_date":"2020-10-06","oldest_available_date":"2000-06-07","column_names":["Date","Open","High","Low","Close","Change","Traded Volume","Turnover","Last Price of the Day","Daily Traded Units","Daily Turnover"],"frequency":"daily","type":"Time Series","premium":false,"limit":null,"transform":null,"column_index":null,"start_date":"2020-09-30","end_date":"2020-09-30","data":[["2020-09-30",108.9,109.0,106.5,108.0,null,208271.0,22455777.3,null,null,null]],"collapse":null,"order":null,"database_id":6129}}


These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [6]:
# TASK 1: Collect data from the Franfurt Stock Exchange
url = 'https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?start_date=2017-01-01&end_date=2017-12-31&api_key=t8sGXjPhKLC1vLVUsUsn' 
r = requests.get(url)

In [7]:
# TASK 2: Convert the returned JSON object into a Python dictionary
json_data = r.json()

# confirm object in type dict
print(type(json_data)) 

<class 'dict'>


In [8]:
# Task 3: Calculate what the highest and lowest opening prices were for the stock in this period

# find which value in the key 'column_names' contains info on 'Open' prices.
col_ = json_data['dataset']['column_names']
print(col_)

['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover']


Column 2 in this dict contains the information on pricing needed. 

In [9]:
# Isolate dict "data"
data_ = json_data['dataset']['data']

In [10]:
# Get'Open' data (2nd item) from every list in 'data_'
def Extract_Open(data_):
    return [item[1] for item in data_]

openData = Extract_Open(data_)

# Remove None from new list
openData_NA = [i for i in openData if i]
print(openData_NA)

[51.76, 51.65, 51.45, 51.05, 51.16, 51.88, 52.73, 52.37, 52.7, 53.11, 52.64, 52.29, 52.28, 51.5, 50.89, 50.8, 51.21, 49.5, 49.52, 48.64, 49.64, 49.09, 49.13, 49.11, 48.8, 48.4, 47.25, 46.57, 47.03, 47.09, 47.98, 48.4, 48.38, 47.3, 47.65, 46.42, 46.16, 45.81, 45.0, 45.88, 46.29, 46.53, 45.48, 45.2, 45.01, 45.16, 44.9, 45.08, 45.72, 46.01, 45.8, 45.61, 45.5, 45.58, 45.97, 45.64, 46.2, 46.19, 46.01, 45.36, 44.51, 43.58, 42.0, 42.35, 42.3, 42.3, 41.48, 42.29, 42.54, 42.65, 42.5, 42.29, 42.35, 42.49, 43.21, 42.81, 42.7, 43.0, 42.66, 43.0, 42.38, 42.16, 42.0, 42.0, 41.71, 42.11, 42.64, 42.72, 42.82, 42.46, 42.42, 42.28, 41.88, 42.4, 42.53, 42.12, 41.3, 41.73, 43.5, 44.9, 45.85, 45.13, 45.34, 45.25, 45.24, 44.94, 45.26, 45.16, 44.91, 44.7, 45.31, 45.57, 45.74, 45.06, 45.5, 45.6, 45.07, 44.67, 44.29, 44.94, 44.64, 44.79, 45.5, 44.67, 45.83, 45.29, 45.01, 45.73, 46.68, 47.23, 46.95, 47.29, 47.03, 47.46, 46.48, 46.9, 45.66, 46.34, 46.52, 46.5, 47.31, 46.77, 47.8, 47.01, 47.12, 46.8, 46.12, 45.22

In [11]:
# sort'OpenData' in ascending order to see highest and lowest opening prices
openData_NA.sort(reverse=True) 
print(openData_NA)

[53.11, 52.73, 52.7, 52.64, 52.37, 52.29, 52.28, 51.88, 51.76, 51.65, 51.5, 51.45, 51.21, 51.16, 51.05, 50.89, 50.8, 49.64, 49.52, 49.5, 49.13, 49.11, 49.09, 48.8, 48.64, 48.4, 48.4, 48.38, 47.98, 47.8, 47.65, 47.46, 47.31, 47.3, 47.29, 47.25, 47.23, 47.12, 47.09, 47.03, 47.03, 47.01, 46.95, 46.9, 46.8, 46.77, 46.68, 46.57, 46.53, 46.52, 46.5, 46.48, 46.42, 46.34, 46.29, 46.2, 46.19, 46.16, 46.12, 46.01, 46.01, 45.97, 45.88, 45.85, 45.83, 45.81, 45.8, 45.74, 45.73, 45.72, 45.66, 45.64, 45.61, 45.61, 45.6, 45.58, 45.57, 45.5, 45.5, 45.5, 45.48, 45.36, 45.34, 45.31, 45.29, 45.26, 45.25, 45.24, 45.22, 45.2, 45.18, 45.16, 45.16, 45.15, 45.13, 45.09, 45.08, 45.07, 45.06, 45.06, 45.05, 45.01, 45.01, 45.0, 44.94, 44.94, 44.91, 44.9, 44.9, 44.8, 44.8, 44.79, 44.7, 44.67, 44.67, 44.64, 44.51, 44.29, 44.16, 44.0, 43.92, 43.74, 43.67, 43.58, 43.5, 43.5, 43.4, 43.21, 43.0, 43.0, 43.0, 42.82, 42.81, 42.72, 42.7, 42.66, 42.65, 42.64, 42.54, 42.53, 42.52, 42.5, 42.5, 42.49, 42.46, 42.42, 42.4, 42.38,

The highest and lowest opening prices were for the stock in this period are 53.11 and 34.0 respectively.

In [12]:
# Task 4: The largest change in any one day (based on High and Low price)

# Get'high'[2] data from every list in 'data_'
def Extract_high(data_):
    return [item[2] for item in data_]

highData = Extract_high(data_)

# Remove None from new list
highData_NA = [i for i in highData if i]
print(highData_NA)

[51.94, 51.82, 51.89, 51.5, 51.52, 52.04, 52.73, 52.75, 52.7, 53.54, 53.35, 53.1, 52.45, 52.83, 51.47, 51.11, 51.38, 51.23, 50.49, 49.84, 49.64, 49.89, 49.73, 49.41, 49.46, 49.61, 48.59, 47.38, 47.15, 47.23, 48.01, 48.9, 48.61, 48.89, 48.06, 47.72, 46.33, 46.09, 46.04, 46.06, 46.55, 46.65, 46.42, 45.41, 45.06, 45.27, 45.34, 45.34, 45.85, 46.2, 46.06, 45.75, 45.7, 45.58, 45.97, 46.04, 46.2, 46.19, 46.09, 46.17, 44.98, 44.17, 43.56, 42.49, 42.57, 42.3, 42.38, 42.29, 42.54, 42.65, 42.63, 42.81, 42.8, 42.69, 43.34, 42.89, 42.75, 43.02, 42.71, 43.19, 42.75, 43.06, 42.08, 42.2, 41.98, 42.25, 42.64, 43.05, 43.17, 42.96, 42.76, 42.6, 43.01, 42.62, 42.53, 42.69, 41.94, 41.99, 43.5, 45.09, 46.34, 45.13, 45.54, 45.77, 45.54, 45.75, 45.29, 45.45, 45.33, 45.04, 45.31, 45.88, 45.96, 45.72, 45.55, 46.23, 45.56, 45.18, 45.05, 44.94, 45.18, 44.79, 45.5, 45.36, 45.83, 45.83, 45.74, 45.81, 46.68, 47.33, 47.63, 47.4, 47.4, 47.48, 47.43, 46.9, 46.8, 46.34, 46.86, 46.51, 47.43, 47.44, 47.8, 47.43, 47.45, 46

In [13]:
# Get'low'[3] data from every list in 'data_'
def Extract_low(data_):
    return [item[3] for item in data_]

lowData = Extract_low(data_)

# Remove None from new list
lowData_NA = [i for i in lowData if i]
print(lowData_NA)

[51.45, 51.43, 50.76, 50.92, 50.9, 51.2, 51.07, 51.61, 51.64, 52.15, 52.48, 51.82, 51.26, 51.28, 50.81, 50.39, 50.4, 49.5, 49.17, 48.28, 48.7, 49.03, 48.96, 48.87, 48.45, 48.39, 46.78, 46.54, 46.8, 46.55, 46.75, 47.84, 47.76, 47.16, 47.09, 46.42, 45.84, 45.76, 44.83, 45.18, 45.97, 45.61, 45.46, 44.91, 44.7, 44.75, 44.89, 44.76, 44.79, 45.61, 45.37, 45.3, 45.37, 45.17, 45.25, 45.57, 45.6, 45.69, 45.63, 45.22, 44.18, 43.3, 42.0, 41.78, 42.11, 41.96, 41.48, 41.39, 41.99, 42.13, 42.23, 42.25, 42.35, 42.22, 42.62, 42.56, 42.56, 42.67, 42.34, 42.55, 41.95, 42.07, 41.12, 41.49, 41.33, 41.86, 42.05, 42.63, 42.6, 42.4, 42.2, 42.01, 41.76, 41.98, 42.2, 42.01, 40.96, 41.14, 41.64, 44.15, 44.02, 44.36, 44.91, 44.9, 45.1, 44.94, 44.75, 45.15, 44.46, 44.63, 44.49, 45.04, 45.23, 44.94, 44.7, 45.29, 44.83, 44.67, 43.89, 44.08, 44.51, 44.25, 44.15, 44.44, 44.74, 45.06, 45.0, 45.11, 45.41, 46.39, 46.91, 46.79, 46.75, 46.53, 46.27, 46.25, 45.66, 45.21, 46.05, 46.03, 45.89, 46.55, 46.27, 47.01, 46.21, 46.

In [14]:
# Find difference (high - low) of each day
import operator
diff = list(map(operator.sub, highData_NA,lowData_NA))
diff.sort(reverse=True)
print(diff)

[2.8100000000000023, 2.3200000000000003, 2.009999999999998, 1.8599999999999994, 1.8100000000000023, 1.730000000000004, 1.7299999999999969, 1.7299999999999969, 1.6599999999999966, 1.5600000000000023, 1.5600000000000023, 1.5499999999999972, 1.5399999999999991, 1.529999999999994, 1.490000000000002, 1.4699999999999989, 1.4000000000000057, 1.3900000000000006, 1.3500000000000014, 1.3300000000000054, 1.3200000000000003, 1.2999999999999972, 1.2800000000000011, 1.2700000000000031, 1.259999999999998, 1.25, 1.25, 1.240000000000002, 1.230000000000004, 1.2199999999999989, 1.2100000000000009, 1.1900000000000048, 1.1599999999999966, 1.1599999999999966, 1.1599999999999966, 1.1400000000000006, 1.1400000000000006, 1.1300000000000026, 1.1300000000000026, 1.1300000000000026, 1.1199999999999974, 1.0899999999999963, 1.0700000000000003, 1.0700000000000003, 1.0600000000000023, 1.0600000000000023, 1.0599999999999952, 1.0399999999999991, 1.0399999999999991, 1.0200000000000031, 1.009999999999998, 1.0099999999999

The largest difference in a day was 2.81 based on 'High' and 'Low' prices.

In [15]:
# Task 5: The largest change between any two days (based on Closing Price)

# Get'closing'[4] data from every list in 'data_'
def Extract_closing(data_):
    return [item[4] for item in data_]

closingData = Extract_closing(data_)

# Remove None from new list
closingData_NA = [i for i in closingData if i]
print(closingData_NA)

[51.76, 51.6, 51.82, 51.32, 51.4, 51.27, 51.66, 52.62, 52.01, 52.67, 53.09, 52.43, 52.14, 52.12, 51.47, 50.89, 51.25, 51.14, 49.86, 49.7, 48.75, 49.25, 49.2, 49.11, 49.2, 48.8, 48.39, 47.04, 46.84, 47.03, 47.05, 48.0, 48.34, 48.34, 47.21, 47.47, 46.26, 45.99, 45.97, 45.27, 46.04, 45.76, 46.41, 45.41, 45.0, 44.85, 45.0, 44.87, 45.0, 45.77, 45.96, 45.55, 45.4, 45.43, 45.29, 45.84, 45.74, 46.0, 46.05, 46.11, 44.98, 44.17, 43.56, 42.04, 42.37, 42.07, 42.06, 41.46, 41.99, 42.44, 42.27, 42.42, 42.52, 42.45, 42.73, 42.85, 42.67, 42.77, 42.55, 42.62, 42.6, 42.41, 41.9, 41.94, 41.85, 41.91, 42.14, 42.69, 42.71, 42.71, 42.26, 42.41, 42.5, 42.05, 42.28, 42.3, 41.94, 41.68, 41.81, 44.37, 44.96, 45.07, 44.97, 45.56, 45.45, 45.3, 44.97, 45.25, 45.16, 44.82, 44.61, 45.44, 45.66, 45.57, 45.0, 45.6, 45.53, 44.95, 44.95, 44.2, 44.7, 44.53, 44.62, 45.19, 44.8, 45.75, 45.44, 45.2, 45.68, 46.83, 47.21, 46.99, 47.29, 46.99, 47.37, 46.64, 46.63, 45.67, 46.33, 46.32, 46.31, 47.44, 46.27, 47.43, 47.43, 46.99, 

In [16]:
# Calculate difference list
diff_list = []
for x, y in zip(closingData_NA[0::], closingData_NA[1::]):
    diff_list.append(y-x)
    diff_list.sort(reverse=True)
print(diff_list)

[2.559999999999995, 2.039999999999999, 1.1599999999999966, 1.1499999999999986, 1.1299999999999955, 1.1200000000000045, 0.990000000000002, 0.9600000000000009, 0.9500000000000028, 0.9500000000000028, 0.8299999999999983, 0.7700000000000031, 0.769999999999996, 0.6799999999999997, 0.6600000000000037, 0.6599999999999966, 0.6499999999999986, 0.6099999999999994, 0.6000000000000014, 0.5900000000000034, 0.5900000000000034, 0.5899999999999963, 0.5700000000000003, 0.5500000000000043, 0.5499999999999972, 0.5300000000000011, 0.5200000000000031, 0.509999999999998, 0.5, 0.5, 0.4799999999999969, 0.46000000000000085, 0.44999999999999574, 0.4399999999999977, 0.4200000000000017, 0.3999999999999986, 0.39000000000000057, 0.38999999999999346, 0.38000000000000256, 0.37999999999999545, 0.36999999999999744, 0.35999999999999943, 0.3400000000000034, 0.3399999999999963, 0.3299999999999983, 0.3299999999999983, 0.3200000000000003, 0.29999999999999716, 0.28999999999999915, 0.28000000000000114, 0.28000000000000114, 0.

The largest change between any two days was 2.56 based on Closing Price.

In [17]:
# Task 6: The average daily trading volume during this year

# Get'Traded Volume'[6] data from every list in 'data_'
def Extract_volume(data_):
    return [item[6] for item in data_]

volumeData = Extract_volume(data_)

# Remove None from new list
volumeData_NA = [i for i in volumeData if i]
print(volumeData_NA)

[34640.0, 40660.0, 57452.0, 71165.0, 120649.0, 50587.0, 137313.0, 129733.0, 204080.0, 132981.0, 128434.0, 87911.0, 71817.0, 109157.0, 48123.0, 88730.0, 83023.0, 94385.0, 101733.0, 123019.0, 67342.0, 42669.0, 102180.0, 50350.0, 38834.0, 91142.0, 78502.0, 97252.0, 54107.0, 89373.0, 67593.0, 67672.0, 76286.0, 90245.0, 120268.0, 94195.0, 48152.0, 60716.0, 56911.0, 37958.0, 56319.0, 56245.0, 74472.0, 56319.0, 47730.0, 43042.0, 43375.0, 55707.0, 59991.0, 45263.0, 65837.0, 49246.0, 43362.0, 49180.0, 69455.0, 65860.0, 44059.0, 66760.0, 94804.0, 115706.0, 95313.0, 99821.0, 157234.0, 76600.0, 51321.0, 56224.0, 79955.0, 105194.0, 57838.0, 65546.0, 44037.0, 107144.0, 65157.0, 68801.0, 52828.0, 103273.0, 59881.0, 64320.0, 71006.0, 66351.0, 105288.0, 151474.0, 157888.0, 97804.0, 98156.0, 47130.0, 69734.0, 65213.0, 70269.0, 95376.0, 68812.0, 72886.0, 131361.0, 104676.0, 64334.0, 127682.0, 183412.0, 175161.0, 355857.0, 156168.0, 164543.0, 96202.0, 77854.0, 187468.0, 74975.0, 62672.0, 114006.0, 50557.0

In [18]:
# Calcuate the daily average:
len(volumeData_NA) # num of days recorded = 255
avg = sum(volumeData_NA)/255
print(round(avg,2))

89124.34


The average daily trading volume during this year was 89,124.34.

In [19]:
# Task 7: The median trading volume during this year
import statistics
statistics.median(volumeData_NA)

76286.0

The median trading volume during this year was 76,286.